niworkflows/viz/plots.py

```diff
@@ -271,6 +290,128 @@ def plot_carpet(img, atlaslabels, detrend=True, nskip=0, size=(950, 800),
+    # Define TR and number of frames
+    notr = False
+    if tr is None:
+        notr = True
+        tr = 1.
+
+    data = cimg.get_fdata().T
+    struct_map = {'LEFT_CORTEX': 1, 'RIGHT_CORTEX': 2, 'SUBCORTICAL': 3, 'CEREBELLUM': 4}
+    seg = np.zeros((data.shape[0], ), dtype='uint32')
+    for bm in matrix.get_index_map(1).brain_models:
+        if 'CORTEX' in bm.brain_structure:
+            lidx = (1, 2)['RIGHT' in bm.brain_structure]
+        elif 'CEREBELLUM' in bm.brain_structure:
+            lidx = 4
+        else:
+            lidx = 3
+        index_final = bm.index_offset + bm.index_count
+        seg[bm.index_offset:index_final] = lidx
+    assert len(seg[seg < 1]) == 0, "Unassigned labels"
+
+    # reorder preserving as much continuity as possible
+    order = seg.argsort(kind='stable')
+
+    # Detrend data
+    v = (None, None)
+    if detrend:
+        data = clean(data.T, t_r=tr).T
+        v = (-2, 2)
+
+    # If subplot is not defined
+    if subplot is None:
+        subplot = mgs.GridSpec(1, 1)[0]
+
+    # Define nested GridSpec
+    wratios = [1, 100, 20]
+    gs = mgs.GridSpecFromSubplotSpec(1, 2, subplot_spec=subplot,
+                                     width_ratios=wratios[:2],
+                                     wspace=0.0)
+
+    mycolors = [cm.get_cmap('Paired').colors[i] for i in (1, 0, 7, 3)]
+    cmap = ListedColormap(mycolors)
+    assert len(cmap.colors) == len(struct_map), \
+        "Mismatch between expected # of structures and colors"
+
+    # Segmentation colorbar
+    ax0 = plt.subplot(gs[0])
+    ax0.set_yticks([])
+    ax0.set_xticks([])
+    ax0.imshow(seg[order, np.newaxis], interpolation='none', aspect='auto', cmap=cmap)
+    ax0.grid(False)
+    ax0.spines["left"].set_visible(False)
+    ax0.spines["bottom"].set_color('none')
+    ax0.spines["bottom"].set_visible(False)
+
+    # Carpet plot
+    ax1 = plt.subplot(gs[1])
+    ax1.imshow(data[order], interpolation='nearest', aspect='auto', cmap='gray',
+               vmin=v[0], vmax=v[1])
+    ax1.grid(False)
+    ax1.set_yticks([])
+    ax1.set_yticklabels([])
+
+    # Set 10 frame markers in X axis
+    interval = max((int(data.shape[-1] + 1) // 10, int(data.shape[-1] + 1) // 5, 1))
+    xticks = list(range(0, data.shape[-1])[::interval])
+    ax1.set_xticks(xticks)
+    ax1.set_xlabel('time (frame #)' if notr else 'time (s)')
+    labels = tr * (np.array(xticks))
+    ax1.set_xticklabels(['%.02f' % t for t in labels.tolist()], fontsize=5)
+
+    # Remove and redefine spines
+    for side in ["top", "right"]:
+        # Toggle the spine objects
+        ax0.spines[side].set_color('none')
+        ax0.spines[side].set_visible(False)
+        ax1.spines[side].set_color('none')
+        ax1.spines[side].set_visible(False)
+
+    ax1.yaxis.set_ticks_position('left')
+    ax1.xaxis.set_ticks_position('bottom')
+    ax1.spines["bottom"].set_visible(False)
+    ax1.spines["left"].set_color('none')
+    ax1.spines["left"].set_visible(False)
```

Comment:
I don't love all this duplicated code... Would it be possible to extract this out to a common, private function for both volumetric and cifti2 carpets? In the end the pseudocode of the plotting would be:

```Python
def _carpet(data, segments, lut):
    <the common code for plotting>
    return plot

def plot_carpet(data, datatype='NIFTI'):
    if datatype.upper().startswith("cifti"):
        """calculate segments and color lookup table (lut) for CIFTI"""
    else:
        """calculate segments and color lookup table (lut) for NIfTI"""
    
    return _carpet(data, segments, lut)